core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+ * under the License.
+ */
+
+package org.elasticsearch.threadpool;
+
+import org.elasticsearch.common.collect.HppcMaps.Object;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.BaseFuture;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.ThreadPool.Cancellable;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+import org.elasticsearch.threadpool.ThreadPool.ReschedulingRunnable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.sameInstance;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.verifyZeroInteractions;
+
+/**
+ * Unit tests for the scheduling of tasks with a fixed delay
+ */
+public class ScheduleWithFixedDelayTests extends ESTestCase {
+
+    public void testThatRunnableIsRescheduled() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(scaledRandomIntBetween(2, 16));
+        final Runnable countingRunnable = latch::countDown;
+
+        ThreadPool threadPool = new ThreadPool("counting-runnable");
+        try {
+            Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(100L), Names.GENERIC);
+            assertNotNull(cancellable);
+            latch.await();
+            cancellable.cancel();
+            assertTrue(cancellable.isCancelled());
+        } finally {
+            threadPool.shutdown();
+        }
+    }
+
+    public void testCancellingRunnable() throws Exception {
+        final int valueToWaitFor = scaledRandomIntBetween(2, 16);
+        final AtomicInteger counter = new AtomicInteger(0);
+        final Runnable countingRunnable = counter::incrementAndGet;
+
+        ThreadPool threadPool = new ThreadPool("cancelling-runnable");
+        try {
+            Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(50), Names.GENERIC);
+            final boolean counterMetExpected = awaitBusy(() -> counter.get() >= valueToWaitFor, 2L, TimeUnit.SECONDS);
+            assertTrue(counterMetExpected);
+
+            cancellable.cancel();
+            final int currentCount = counter.get();
+            assertTrue(cancellable.isCancelled());
+
+            // we cancel but there could be a currently running iteration, so we buffer by one value...
+            final boolean stillRunning = awaitBusy(() -> counter.get() > currentCount + 1, 250L, TimeUnit.MILLISECONDS);
```

Comment:
yeah, this is tricky. I instead I think we can run a test that cancels from the running task and makes sure it is never run again? we can then decide whether to cancel immediately or after a few iterations